lib/mix/lib/mix/local/fetcher.ex

```diff
@@ -0,0 +1,70 @@
+defmodule Mix.Local.Fetcher do
+  @moduledoc false
+
+  @doc """
+  Generates a new mix project in a temporary directory with the given
+  `package_name` and `dep_spec` added to a mix.exs. Then, `in_fetcher` is
+  executed in the fetcher project. By default, this fetches the dependencies,
+  but you can provide an `in_fetcher` during test or for other purposes. After
+  the `in_fetcher` is executed, `in_package` is executed in the now (presumably)
+  fetched package, with the package's config overridden with the deps_path and
+  lockfile of the fetcher package. Also, the Mix env is set to :prod.
+  """
+  @spec fetch(String.t, tuple, ((atom) -> any), ((atom) -> any)) :: any
+  def fetch(package_name, dep_spec, in_fetcher \\ &in_fetcher/1, in_package) do
```

Comment:
This looks great! I think we should just add it to Mix.Local.Installer though, specially because it is very likely all the remaining functionality will be there anyway... although we can wait to move this code if you want until we support archives.